### PR TITLE
Add --stale-worktree-timeout option

### PR DESCRIPTION
**Pull request by sviscaino**
Adds an option to wait for a certain duration before cleaning up the old worktrees.

Test:

https://user-images.githubusercontent.com/23248585/235175534-240fc875-5931-4f6f-886b-4624ec69989d.mp4

Resolves https://github.com/kubernetes/git-sync/issues/692

### CODE REVIEW

**Review by thockin — COMMENTED** [review]
This looks pretty nice!  Unfortunately I have a bunch of PRs queued up, so one of us will have to resolve conflicts.  I'll get this in ASAP.

One thing that is missing is what happens if git-sync crashes and restarts.  We still want to clean up.  To do that, we need the timestamp to survive app restarts.  What if we use the worktree's mtime?  When we flip the link, if we don't delete the old wortree, instead touch it.  Then when we do cleanup (my PRs change this area), we read the mtime and only clean up if the mtime is > timeout.

Then you don't need the map at all.

The tests need to go into test_e2e.sh.  Specifically:
* Prove it works for 3 stale worktrees with a 3 second timeout
* Prove it works across restarts (using --run-once)

File: cmd/git-sync/main.go [hunk 1]
```diff
@@ -2074,6 +2096,10 @@ OPTIONS
             Enable the pprof debug endpoints on git-sync's HTTP endpoint.
             Requires --http-bind to be specified.
 
+    --keep-worktrees-duration, $GIT_SYNC_KEEP_WORKTREES_DURATION
+            Keep the past worktrees for the given number of seconds before
```
[the remark this answers is not in the document]
Comment (sviscaino) [reply cmd/git-sync/main.go]: Yeah that's a better description, thanks - made the change

**Review by thockin — COMMENTED** [review]
Ping me when rebased (sorry) and updated.  Looking good.

File: cmd/git-sync/main.go [hunk 2]
```diff
@@ -1272,6 +1299,15 @@ func removeDirContents(dir string, log *logging.Logger, except ...string) error
 			continue
 		}
 		p := filepath.Join(dir, name)
+		if exceptMoreRecentThan > 0 {
+			stat, err := os.Stat(p)
+			if err != nil {
+				log.Error(err, "Unable to stat path", "path", p)
+			}
+			if stat.ModTime().Add(exceptMoreRecentThan).After(time.Now()) {
```
Comment (thockin) [inline cmd/git-sync/main.go:1307]: Can this be rephrased as `time.Since(stat.ModTime()) > exceptMoreRecentThan`

File: cmd/git-sync/main.go [hunk 3]
```diff
@@ -1254,6 +1277,10 @@ func dirIsEmpty(dir string) (bool, error) {
 // removeDirContents iterated the specified dir and removes all contents,
 // except entries which are specifically excepted.
 func removeDirContents(dir string, log *logging.Logger, except ...string) error {
+	return removeDirContentsExceptMostRecent(dir, log, 0, except...)
+}
+
+func removeDirContentsExceptMostRecent(dir string, log *logging.Logger, exceptMoreRecentThan time.Duration, except ...string) error {
```
Comment (thockin) [inline cmd/git-sync/main.go:1283]: How about a slightly different factoring.  I'm thinking something like:

```
func removeDirContentsIf(dir string, log *logging.Logger, fn func(fi os.FileInfo,) (bool, error) ) error {
```

That would read each dirent, call `Stat`, and call the `fn`.  If `fn` returns error, it's an error.  If it returns true, remove the path.

That way, `removeDirContents` just passes a nil func, and the cleanup path can return a local func which compares the timestamps and filenames.  

I *think* this gets is better locality of logic, what do you think?

File: cmd/git-sync/main.go [hunk 4]
```diff
@@ -1692,6 +1728,14 @@ func (git *repoSync) SyncRepo(ctx context.Context, refreshCreds func(context.Con
 		if err != nil {
 			return false, "", err
 		}
+		if currentWorktree != "" {
+			// Touch the old worktree -- which will make cleanupStaleWorktrees delete it in the future,
+			// if the stale timeout option is enabled
+			err = os.Chtimes(currentWorktree.Path(), time.Now(), time.Now())
```
Comment (thockin) [inline cmd/git-sync/main.go:1734]: We could encapsulate this into `touch()` so it's like:

```
err = chtimes
if err is NotFound error
    os.Create()
```

File: test_e2e.sh [hunk 5]
```diff
@@ -576,6 +576,238 @@ function e2e::worktree_cleanup() {
...
+    assert_file_absent "$ROOT"/.worktrees/$WT1/file2
+    assert_file_absent "$ROOT"/.worktrees/$WT1/file3
+    assert_file_exists "$ROOT"/.worktrees/$WT2/file
+    assert_file_exists "$ROOT"/.worktrees/$WT2/file2
+    assert_file_absent "$ROOT"/.worktrees/$WT2/file3
+    assert_file_exists "$ROOT"/.worktrees/$WT3/file
+    assert_file_exists "$ROOT"/.worktrees/$WT3/file2
+    assert_file_exists "$ROOT"/.worktrees/$WT3/file3
+
+    # wait for 4 seconds
```
Comment (thockin) [inline test_e2e.sh:642]: A better comment would be "why" - Wait until the oldest WT expires

File: test_e2e.sh [hunk 6]
```diff
@@ -576,6 +576,238 @@ function e2e::worktree_cleanup() {
...
+    assert_file_absent "$ROOT"/.worktrees/$WT1/file2
+    assert_file_absent "$ROOT"/.worktrees/$WT1/file3
+    assert_file_exists "$ROOT"/.worktrees/$WT2/file
+    assert_file_exists "$ROOT"/.worktrees/$WT2/file2
+    assert_file_absent "$ROOT"/.worktrees/$WT2/file3
+    assert_file_exists "$ROOT"/.worktrees/$WT3/file
+    assert_file_exists "$ROOT"/.worktrees/$WT3/file2
+    assert_file_exists "$ROOT"/.worktrees/$WT3/file3
+
+    # wait for another 2 seconds
```
Comment (thockin) [inline test_e2e.sh:661]: wait for another WT to expire

File: test_e2e.sh [hunk 7]
```diff
@@ -576,6 +576,238 @@ function e2e::worktree_cleanup() {
...
+function e2e::stale_worktree_timeout_restart() {
+    echo "$FUNCNAME 1" > "$REPO"/file
+    git -C "$REPO" commit -qam "$FUNCNAME"
+    WT1=$(git -C "$REPO" rev-list -n1 HEAD)
+    GIT_SYNC \
+        --period=100ms \
+        --repo="file://$REPO" \
+        --root="$ROOT" \
+        --link="link" \
+        --stale-worktree-timeout="5s" \
```
Comment (thockin) [inline test_e2e.sh:693]: don't need this arg on the first run

File: test_e2e.sh [hunk 8]
```diff
@@ -576,6 +576,238 @@ function e2e::worktree_cleanup() {
...
+        --one-time
+
+    # wait for first sync
+    wait_for_sync "${MAXWAIT}"
```
Comment (thockin) [inline test_e2e.sh:697]: no need to wait for sync - the command above won't exit until it is synced

**Review by thockin — COMMENTED** [review]
I'm afraid the timing on the sleeps will be flaky, espcially in the second test - did you see it flaking?  E.g. if docker takes just a little to long to start the next run?

File: cmd/git-sync/main.go (hunk 3 above)
Comment (sviscaino) [reply cmd/git-sync/main.go]: I agree it seems cleaner, I applied your suggestion

**Comment by sviscaino** [comment]
Rebased and comments addressed :)

**Comment by sviscaino** [comment]
> I'm afraid the timing on the sleeps will be flaky, espcially in the second test - did you see it flaking?  E.g. if docker takes just a little to long to start the next run?

I hadn't observed any flakiness locally, but I doubled the delays just in case

File: cmd/git-sync/main.go [hunk 9]
```diff
@@ -475,19 +480,20 @@ func (abs absPath) Base() string {
...
 // repoSync represents the remote repo and the local sync of it.
 type repoSync struct {
-	cmd        string         // the git command to run
-	root       absPath        // absolute path to the root directory
-	repo       string         // remote repo to sync
-	ref        string         // the ref to sync
-	depth      int            // for shallow sync
-	submodules submodulesMode // how to handle submodules
-	gc         gcMode         // garbage collection
-	link       absPath        // absolute path to the symlink to publish
```
Comment (thockin) [inline cmd/git-sync/main.go:485]: was this a merge mistake?

Comment (thockin) [reply cmd/git-sync/main.go:485]: I think this is the only remaining - looks great otherwise.  Thanks for the PR!

Comment (sviscaino) [reply cmd/git-sync/main.go:485]: I just added `staleTimeout time.Duration` at the end, and re-aligned the spacing of the types.

Comment (sviscaino) [reply cmd/git-sync/main.go:485]: Oops, fixed!

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Pull request merged**